File: VerbumKit/Sources/Features/Scripture/ChapterNavigation.swift
import Models

/// Chapter arithmetic over the canon: stepping past the last chapter of a
/// book continues into the next book, and before chapter 1 into the previous
/// book's last chapter. Pure, so the reducer stays small and this stays tested.
enum ChapterNavigation {
    static func next(after reference: PassageReference) -> PassageReference? {
        guard let book = BibleBook.book(id: reference.bookId) else { return nil }
        if reference.chapter < book.chapterCount {
            return PassageReference(bookId: book.id, chapter: reference.chapter + 1)
        }
        guard let nextBook = BibleBook.canon.first(where: { $0.order == book.order + 1 }) else { return nil }
        return PassageReference(bookId: nextBook.id, chapter: 1)
    }

    static func previous(before reference: PassageReference) -> PassageReference? {
        guard let book = BibleBook.book(id: reference.bookId) else { return nil }
        if reference.chapter > 1 {
            return PassageReference(bookId: book.id, chapter: reference.chapter - 1)
        }
        guard let previousBook = BibleBook.canon.first(where: { $0.order == book.order - 1 }) else { return nil }
        return PassageReference(bookId: previousBook.id, chapter: previousBook.chapterCount)
    }
}

File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ChapterNavigation.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.models.BibleBook
import com.nexussoft.verbum.models.PassageReference

/** Chapter arithmetic over the canon; steps roll into the next/previous book. Mirrors iOS. */
internal object ChapterNavigation {
    fun next(after: PassageReference): PassageReference? {
        val book = BibleBook.book(after.bookId) ?: return null
        if (after.chapter < book.chapterCount) return PassageReference(book.id, after.chapter + 1)
        val nextBook = BibleBook.canon.firstOrNull { it.order == book.order + 1 } ?: return null
        return PassageReference(nextBook.id, 1)
    }

    fun previous(before: PassageReference): PassageReference? {
        val book = BibleBook.book(before.bookId) ?: return null
        if (before.chapter > 1) return PassageReference(book.id, before.chapter - 1)
        val previousBook = BibleBook.canon.firstOrNull { it.order == book.order - 1 } ?: return null
        return PassageReference(previousBook.id, previousBook.chapterCount)
    }
}
